pkg/stub/stub.go

```diff
@@ -316,10 +316,10 @@ func (stub *stub) Start(ctx context.Context) (retErr error) {
 	stub.Lock()
 	defer stub.Unlock()
 
-	if stub.started {
+	if stub.isStarted() {
 		return fmt.Errorf("stub already started")
 	}
-	stub.started = true
+	stub.doneC = make(chan struct{})
```

Comment:
In Line 382, could we wait for it if `retErr` isn't nil? 
The goroutine uses reference so that new `Start` would refresh `stub.doneC` and we may have two goroutines to close one channel. We can consider to drain the goroutine if `retErr` != nil.

```
go func() {
		stub.srvErrC <- rpcs.Serve(ctx, rpcl)
		close(stub.doneC)
	}()
```

Reply:
Sorry for late reply.

If `Start` returns error, the following goroutine could be still running and it doesn't reach the `close(stub.doneC)`. And then new `stub.Start()` call will create a new `stub.doneC` channel. If the previous goroutine calls the `close(stub.doneC)` after `stub.doneC = make(chan struct{})`, the `stub.doneC` can be closed twice and then panic. 

```go
go func() {
		stub.srvErrC <- rpcs.Serve(ctx, rpcl)
		close(stub.doneC)
}()
```

For the above goroutine, we can consider to pass var instead of using reference directly.

```go
go func(rpcl xxx, doneC chan struct{}, srvErrC chan error) {
       srvErrC <- rpcs.Serve(ctx, rpcl)
       close(doneC)
}(rpcl, stub.doneC, stub.srvErrC)
```

To be honest, it's hard to reproduce without adding failpoint(sleep). However, we should ensure logic is clean and easy to reason.

Reply:
Thanks for your explanation! you are right, there definitely exists a chance that double goroutine closes one `stub.doneC`;
Does define a separate function like `stub.serve(...)` better?

Reply:
> Does define a separate function like stub.serve(...) better?

It sounds good to me. You can also use anonymous function as it is right now.

Reply:
updated.